### PR TITLE
Support multiple unstable attributes on items

### LINKED ISSUE #94770: Allow multiple #[unstable] attributes on one item
We sometimes have things that should be gated by multiple features. E.g. `ScopedJoinHandle::is_running` is part of `scoped_threads`, but also of `thread_is_running`. In these cases we now just apply only one of the `#[unstable]` attributes, and because of that sometimes accidentally stabilize something when only one of the features gets stabilized.

It'd be great if we could add multiple `#[unstable]` attributes to a single item. Allowing multiple `#[stable]` tags isn't a necessity. We could just remove the `#[unstable]` tag if the feature gets stabilized and there are other `#[unstable]` tags left. The most important part is that we don't forget about the other unstable tags and don't stabilize something by accident.

Another option would be multiple `feature =` arguments to `#[unstable]`.

### PR DESCRIPTION
**Pull request by compiler-errors**
This PR started out pretty simple, but ended up being quite a few changes in a lot of different places to support multiple unstable attributes on items.

Please let me know if I should split this up into separate commits. I feel somewhat bad having one mega-commit, but I wrote this kinda all over the place, so I didn't naturally split it up when I was making these changes.

Fixes #94770

cc: @m-ou-se

### CODE REVIEW

**Comment by compiler-errors** [comment]
Hmmmmm. Seems to be an issue with the way that `Encodable` interacts with the custom `ProcMacroData` encoding. It seems we don't store crate metadata for proc macros, so the span stored in the macro's `StabilityLevel` fails to be decoded.

Not exactly sure how to fix this. Any thoughts @Aaron1011 (I think you're familiar with this code, cc #76897)?

**Comment by petrochenkov** [comment]
@cjgillot do you want to review this (given that you are working on https://github.com/rust-lang/rust/pull/93017)?
If not, then feel free to reassign back.
r? @cjgillot

**Review by cjgillot — COMMENTED** [review]
Thanks for the PR @compiler-errors.
I took a brief look, everything looks good.

What is the motivation for this PR?
What does it mean for an item to be multiply unstable?
How should it be handled/diagnosed?
I saw you have a specific handling for const-unstable attributes, could you elaborate on that?

**Comment by m-ou-se** [comment]
> What is the motivation for this PR?

Imagine we have an unstable type called `NonTwoU32` which can never be 2, gated by `non_two_ints` gate, and an unstable method on all the integer types called `is_power_of_three()` (including u32 and NonZeroU32 and NonTwoU32 etc.), gated by the `is_power_of_three` feature gate. Then `NonTwoU32::is_power_of_three()` should be gated by *both* these feature gates. If either of the features is stabilized, its `#[unstable]` tag can be removed from that specific method, while still keeping the other one.

In the past we've had situations where we accidentally stabilized a method like `NonTwoU32::is_power_of_three()` while stabilizing `NonTwoU32`, while we hadn't yet decided if it was a good idea to have a `is_power_of_three()` method on any type yet.

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs [hunk 1]
```diff
@@ -902,54 +901,72 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
...
-                    if self.ccx.is_const_stable_const_fn() || is_intrinsic {
-                        self.check_op(ops::FnCallUnstable(callee, None));
-                        return;
+                    if !bad_gates.is_empty() {
+                        self.check_op(ops::FnCallUnstable(callee, bad_gates));
+                    }
+                } else {
+                    // FIXME(ecstaticmorse); For compatibility, we consider `unstable` callees that
+                    // have no `rustc_const_stable` attributes to be const-unstable as well. This
+                    // should be fixed later.
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_const_eval/src/transform/check_consts/check.rs:955]: I have no idea. @ecstatic-morse do you know if this FIXME is still valid?

**Comment by compiler-errors** [comment]
@rustbot ready

I believe this is ready for a more thorough review

File: src/librustdoc/clean/inline.rs [hunk 2]
```diff
@@ -351,7 +351,8 @@ crate fn build_impl(
             }
 
             if let Some(stab) = tcx.lookup_stability(did) {
-                if stab.level.is_unstable() && stab.feature == sym::rustc_private {
+                if let rustc_attr::Unstable { unstables, .. } = &stab.level
+                  && unstables.iter().any(|u| u.feature == sym::rustc_private) {
```
Comment (camelid) [inline src/librustdoc/clean/inline.rs:355]: Seems it'd be a good idea to have a `has_feature` helper method on `stab.level`.

File: src/librustdoc/lib.rs [hunk 3]
```diff
@@ -40,7 +41,6 @@ extern crate rustc_ast;
 extern crate rustc_ast_lowering;
 extern crate rustc_ast_pretty;
 extern crate rustc_attr;
-extern crate rustc_const_eval;
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/lib.rs:43]: Yeah, I think it should be re-added.

[the remark this answers is not in the document]
Comment (compiler-errors) [reply src/librustdoc/lib.rs:43]: ahh, so I moved `is_unstable_const_fn` (which was exported from `rustc_const_eval` as a function) to `rustc_middle` as a method on `TyCtxt`, so there is no dependency on `rustc_const_eval` anymore.

In fact, `x check rustdoc` fails if I add this crate back:
```
error: unused extern crate
  --> src/librustdoc/lib.rs:44:1
   |
44 | extern crate rustc_const_eval;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove it
```

File: src/librustdoc/html/render/mod.rs [hunk 4]
```diff
@@ -667,25 +667,35 @@ fn short_item_info(
...
     // Those crates are permanently unstable so it makes no sense to render "unstable" everywhere.
-    if let Some((StabilityLevel::Unstable { reason: _, issue, .. }, feature)) = item
-        .stability(cx.tcx())
-        .as_ref()
-        .filter(|stab| stab.feature != sym::rustc_private)
-        .map(|stab| (stab.level, stab.feature))
+    if let Some(rustc_attr::Stability {
+        level: StabilityLevel::Unstable { unstables, .. }, ..
+    }) = item.stability(cx.tcx())
     {
```
Comment (cjgillot) [inline src/librustdoc/html/render/mod.rs:673]: This block will add the "nightly-only experimental API" message even if there is only the `rustc_private` feature. Why this change in behaviour?

File: src/librustdoc/html/render/mod.rs [hunk 5]
```diff
@@ -667,25 +667,35 @@ fn short_item_info(
...
         let mut message =
             "<span class=\"emoji\">🔬</span> This is a nightly-only experimental API.".to_owned();
```
Comment (cjgillot) [inline src/librustdoc/html/render/mod.rs:675]: This string and the parentheses could be incorporated to the `extra_info.push` below.

**Review by cjgillot — COMMENTED** [review]
The implementation looks good overall, left a few nits.

Second question about semantics:
Unstability attributes can be inherited from an enclosing block (examples below). In this implementation, the innermost attribute overrides attributes from outer scopes. I wonder if they should not be made cumulative:
```
#[rustc_unstable(feature="foo")]
struct Foo {
  #[rustc_unstable(feature="bar")]
  Bar()
  // Is this unstable by ["foo", "bar"] or only ["bar"]?
}

#[rustc_const_unstable(feature="foo")]
impl Foo for () {
  #[rustc_const_unstable(feature="bar")]
  fn bar() {}
  // Is this unstable by ["foo", "bar"] or only ["bar"]?
}
```

**Comment by compiler-errors** [comment]
> Second question about semantics

@m-ou-se, do you know if we should change the behavior so that the unstable attributes inherited from parent modules are additive?

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ecstatic-morse) [reply compiler/rustc_const_eval/src/transform/check_consts/check.rs:961]: IIRC, these rules proved somewhat contentious, although I don't know if this particular idea was the problematic one. You can take out the FIXME and the "this should be fixed", but please leave the description if it's still correct.

**Comment by m-ou-se** [comment]
> > Second question about semantics
> 
> @m-ou-se, do you know if we should change the behavior so that the unstable attributes inherited from parent modules are additive?

Oh, that's a good question. I'd expect them to be additive, but we should take a look at the current `#[unstable]` attributes in core/alloc/std, and see what makes sense there. Having an unstable module with stable items (or differently-unstable items) inside it sounds wrong, but I suspect we actually use that in some places in combination with a re-export from a stable module.

**Comment by m-ou-se** [comment]
> @m-ou-se, who would be the right person to check this? I can definitely give it a first look-over, but might need some help from T-libs to know what is okay and what might break due to this PR.

I think @yaahc recently looked into this, but I might be wrong.

**Comment by yaahc** [comment]
> I think @yaahc recently looked into this, but I might be wrong.

You are not wrong! I'm working on something very closely related in https://github.com/rust-lang/rust/pull/95956 where I'm adding support for checking the stability of path segments when naming items. This in effect makes them additive and would end up duplicating the functionality in this PR if you added parent module additive inheritance.

> Having an unstable module with stable items (or differently-unstable items) inside it sounds wrong, but I suspect we actually use that in some places in combination with a re-export from a stable module.

Like you said, this seems wrong but we already do this in multiple places, between `core::unicode::UNICODE_VERSION` and many of the intrinsics like `core::intrinsics::transmute`. Not to mention that I'm looking to depend upon such functionality to move `Error` into `core` unstably, by having an unstable `core::error` module with a stable `Error` trait in it that is then re-exported from the stable `std::error` module. I think it's important that we preserve that functionality.

Based on that and how hard I anticipate auditing the inherited stability attributes to be, I think the right call is to not make them inherit additively in this PR, and instead rely on the stable-in-unstable PR's path segment stability checks to get the additive functionality between parent module stability and item stability.

File: src/librustdoc/html/render/mod.rs [hunk 6]
```diff
@@ -920,16 +930,17 @@ fn render_stability_since_raw(
...
-            if Some(since) != containing_const_ver =>
+        Some(ConstStability { level: StabilityLevel::Stable { since, .. }, .. })
+            if Some(*since) != containing_const_ver =>
         {
             Some((format!("const since {}", since), format!("const: {}", since)))
         }
-        Some(ConstStability { level: StabilityLevel::Unstable { issue, .. }, feature, .. }) => {
-            let unstable = if let Some(n) = issue {
+        Some(ConstStability { level: StabilityLevel::Unstable { unstables, .. }, .. }) => {
+            // FIXME(compiler-errors): Should we link to several issues here? How?
```
[the remark this answers is not in the document]
Comment (jsha) [reply src/librustdoc/html/render/mod.rs:939]: The current style is like:

const: [unstable]() · [source]()

I don't think there's room in there for multiple unstable links. Instead I'd recommend picking the first one to link to, and solving this in the issue descriptions by cross-linking to issues that apply to intersection sets of items.

**Comment by compiler-errors** [comment]
This has PR has rotted so much that it probably needs to be re-done from scratch. I'll close this and unclaim the issue.
